### PR TITLE
conformance against invalid ReferenceGrants in HTTPRoute and TLSRoute

### PR DESCRIPTION
**Pull request by meyskens**
**What type of PR is this?**
/kind test
/area conformance

**What this PR does / why we need it**:

This PR adds two new conformance tests to test that invalid ReferenceGrants are not getting programmed as well as checking that they are indeed not. It does this check for both HTTPRoutes and TLSRoutes. The list of invalid ReferenceGrants is based on the test in https://github.com/kubernetes-sigs/gateway-api/blob/main/conformance/tests/gateway-secret-invalid-reference-grant.go

**Which issue(s) this PR fixes**:

Fixes https://github.com/kubernetes-sigs/gateway-api/issues/2075

**Does this PR introduce a user-facing change?**:
```release-note
Add conformance tests against accepting invalid ReferenceGrants in HTTPRoute and TLSRoute
```

### CODE REVIEW

File: conformance/tests/httproute-invalid-reference-grant.yaml [hunk 1]
```diff
@@ -0,0 +1,119 @@
...
+  name: reference-grant-wrong-to-kind
+  namespace: gateway-conformance-web-backend
+spec:
+  from:
+    - group: gateway.networking.k8s.io
+      kind: HTTPRoute
+      namespace: gateway-conformance-infra
+  to:
+    - group: ""
+      kind: Service
```
Comment (sunjayBhatia) [inline conformance/tests/httproute-invalid-reference-grant.yaml:89]: seems like this should be Secret and the below Service

File: conformance/tests/httproute-invalid-reference-grant.yaml [hunk 2]
```diff
@@ -0,0 +1,119 @@
...
+  namespace: gateway-conformance-web-backend
+spec:
+  from:
+    - group: gateway.networking.k8s.io
+      kind: HTTPRoute
+      namespace: gateway-conformance-infra
+  to:
+    - group: not-the-group-youre-looking-for
+      kind: Service
+      name: not-the-service-youre-looking-for
```
[the remark this answers is not in the document]
Comment (meyskens) [reply conformance/tests/httproute-invalid-reference-grant.yaml]: Thanks! fixed both :)

**Review by sunjayBhatia — CHANGES_REQUESTED** [review]
missing from here: https://github.com/kubernetes-sigs/gateway-api/blob/c8b54a05c850cd717eb852c4874c6c89d02a5ef8/conformance/utils/config/timeout.go#L108

```go
	if timeoutConfig.TLSRouteMustHaveCondition == 0 {
		timeoutConfig.TLSRouteMustHaveCondition = defaultTimeoutConfig.TLSRouteMustHaveCondition
	}
```

File: conformance/tests/httproute-invalid-reference-grant.go [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, HTTPRouteInvalidReferenceGrant)
+}
+
+var HTTPRouteInvalidReferenceGrant = suite.ConformanceTest{
+	ShortName:   "HTTPRouteInvalidReferenceGrant",
+	Description: "A single HTTPRoute in the gateway-conformance-infra namespace, with a backendRef in another namespace without valid ReferenceGrant, should not attach to Gateway in the gateway-conformance-infra namespace",
```
Comment (sunjayBhatia) [inline conformance/tests/httproute-invalid-reference-grant.go:37]:
```suggestion
	Description: "A single HTTPRoute in the gateway-conformance-infra namespace, with a backendRef in another namespace without valid ReferenceGrant, should have the ResolvedRefs condition set to False and not forward HTTP requests to any backend",
```

File: conformance/tests/tlsroute-invalid-reference-grant.go [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+	"sigs.k8s.io/gateway-api/conformance/utils/suite"
+)
+
+func init() {
+	ConformanceTests = append(ConformanceTests, TLSRouteInvalidReferenceGrant)
+}
+
+var TLSRouteInvalidReferenceGrant = suite.ConformanceTest{
+	ShortName:   "TLSRouteInvalidReferenceGrant",
+	Description: "A single TLSRoute in the gateway-conformance-infra namespace, with a backendRef in another namespace without valid ReferenceGrant, should not attach to Gateway in the gateway-conformance-infra namespace",
```
Comment (sunjayBhatia) [inline conformance/tests/tlsroute-invalid-reference-grant.go:36]: similar change to the above

File: conformance/base/manifests.yaml [hunk 5]
```diff
@@ -300,6 +300,75 @@ spec:
 ---
 apiVersion: v1
 kind: Namespace
+metadata:
+  name: gateway-conformance-tls-backend
```
Comment (sunjayBhatia) [inline conformance/base/manifests.yaml:304]: could be reuse the existing `-app-backend` instead of having to create a new namespace?

**Comment by sunjayBhatia** [comment]
(with the timeout change, these pass against Contour)

File: conformance/base/manifests.yaml (hunk 5 above)
Comment (meyskens) [reply conformance/base/manifests.yaml:304]: i was doubting to do to that but not entirely too sure about it so I chose to make a separate namespace to make clear boundaries. Happy to revert it  and use the app namespace.

Comment (sunjayBhatia) [reply conformance/base/manifests.yaml:304]: i think one of the maintainers may chime in here but I believe we've tried to limit the number of namespaces/resources the tests create to streamline things a bit 👍🏽

File: conformance/utils/kubernetes/helpers.go [hunk 6]
```diff
@@ -611,6 +610,40 @@ func GatewayAndTLSRoutesMustBeAccepted(t *testing.T, c client.Client, timeoutCon
...
+		err := client.Get(ctx, routeNN, route)
+		if err != nil {
+			return false, fmt.Errorf("error fetching TLSRoute: %w", err)
+		}
+
+		parents := route.Status.Parents
+		var conditionFound bool
+		for _, parent := range parents {
+			if err := ConditionsHaveLatestObservedGeneration(route, parent.Conditions); err != nil {
+
```
Comment (sunjayBhatia) [inline conformance/utils/kubernetes/helpers.go:629]: I believe the lint rules have become more strict so the CI job is failing due to small things like this extra newline

File: conformance/base/manifests.yaml [hunk 7]
```diff
@@ -307,6 +307,75 @@ metadata:
...
+        secret:
+          secretName: tls-passthrough-checks-certificate
+          items:
+          - key: tls.crt
+            path: crt
+          - key: tls.key
+            path: key
+---
+apiVersion: v1
+kind: Namespace
```
Comment (sunjayBhatia) [inline conformance/base/manifests.yaml:371]: looks like this is a duplicate now, I think it can be removed

Comment (meyskens) [reply conformance/base/manifests.yaml:371]: you're right... oops :)

File: conformance/base/manifests.yaml [hunk 8]
```diff
@@ -307,6 +307,68 @@ metadata:
...
+          value: /etc/secret-volume/crt
+        - name: TLS_SERVER_PRIVKEY
+          value: /etc/secret-volume/key
+        resources:
+          requests:
+            cpu: 10m
+      volumes:
+      - name: secret-volume
+        secret:
+          secretName: tls-passthrough-checks-certificate
```
Comment (sunjayBhatia) [inline conformance/base/manifests.yaml:363]: running the tests, looks like this secret is missing from this namespace, i think you're missing creating it in `conformance/suite/suite.go`

File: conformance/tests/httproute-invalid-reference-grant.yaml
[the remark this answers is not in the document]
Comment (robscott) [reply conformance/tests/httproute-invalid-reference-grant.yaml:1]: Yeah this is really great, nicely done @meyskens!

File: conformance/tests/httproute-invalid-reference-grant.go [hunk 9]
```diff
@@ -0,0 +1,71 @@
+/*
+Copyright 2022 The Kubernetes Authors.
```
Comment (robscott) [inline conformance/tests/httproute-invalid-reference-grant.go:2]: Same comment applies for all new go files.
```suggestion
Copyright 2023 The Kubernetes Authors.
```

**Review by robscott — COMMENTED** [review]
Thanks @meyskens! Other than @sunjayBhatia's comment, this mostly LGTM.

File: conformance/base/manifests.yaml (hunk 8 above)
Comment (meyskens) [reply conformance/base/manifests.yaml:363]: Added them there aswell :)

**Review by sunjayBhatia — APPROVED** [review]
LGTM, tests both pass with Contour

one comment here that needs a follow up issue: https://github.com/kubernetes-sigs/gateway-api/pull/2076/files#r1245405876

File: conformance/tests/tlsroute-invalid-reference-grant.yaml [hunk 10]
```diff
@@ -0,0 +1,141 @@
...
+  rules:
+    - backendRefs:
+        - name: tls-backend
+          namespace: gateway-conformance-app-backend
+          port: 443
+---
+apiVersion: gateway.networking.k8s.io/v1beta1
+kind: Gateway
+metadata:
+  name: gateway-tlsroute
```
Comment (sunjayBhatia) [inline conformance/tests/tlsroute-invalid-reference-grant.yaml:126]: one more nit, we may want to make sure the name here and the TLSRoute above do not conflict with the existing TLSRoute test to ensure the tests can be run in parallel

Comment (meyskens) [reply conformance/tests/tlsroute-invalid-reference-grant.yaml:126]: Good point! Changed it :)

**Comment by robscott** [comment]
Thanks @meyskens!

/lgtm
/approve

**Pull request merged**